### PR TITLE
refactor robot framework tests

### PR DESCRIPTION
**Pull request by reubenmiller**
Signed-off-by: Reuben Miller <reuben.d.miller@gmail.com>

## Proposed changes

This PR includes a major refactoring/reorganizing of the Robot Framework tests. The refactoring exercise is not 100% complete, however it at least at a point where it will provide benefits for both testing and development. In addition once it is merged it is then easier to incrementally improve upon the coverage, by porting more PySys tests to Robot Framework.

The goal of this PR is to:

* Enable integration tests to run on PR checks
* Make it easier to write and understand tests
* Make it possible/easy to run tests on virtual devices (e.g. a multi-process container)
* Provide developer friendly tools to make it easier to try out their code on temporary devices
* Make it possible to run (most of) the same tests on different devices (e.g. container or a real device)
* Create external robot framework libraries that can be potentially used by the Community (though the packages are not published to pypi just yet)

**Summary of changes**

* New `integration-tests` workflow which is triggered on PRs. Though it requires a manual confirmation before it is given access to integration test secrets (for security purposes). Tests are only run on `x86_64` target as a first initial sanity/regression tests.
* Move logic out into external libraries (e.g. Cumulocity specific logic, device creation logic etc.)
* Use python virtual environment to install integration test (python) dependencies
* Added docker-compose example and instructions how to quickly create a test device (for manual local testing)

**What is still missing / will be done in follow up PRs**

* Ensuring that the applicable tests can run against a container or a ssh device without rewriting any tests (so far only the container interface has been verified)
* Fix any flakey integration tests (if it arises...this is unfortunately fairly normal for integration tests)
* Add integration tests to the `main` branch and release workflows

### Background

Since this is such a large PR, I wanted to provide more information about the motivation and give more details on some of the changes. These have been grouped into the following sections.

#### Custom robot framework libraries instead of user-created keywords

Some of the test were getting very complex as they relied on having detailed knowledge of the Cumulocity IoT REST interfaces. To simplify the process the complex device and cloud interactions have been moved out into external projects where they will also can be made available for use by other thin-edge.io plugin authors.

The following are small examples of what keywords were created in the new external libraries. For example, for a Cumulocity Alarm assertion, the keyword would just be:

```
Device Should Have Measurements    minimum=1    maximum=1    type=ThinEdgeMeasurement
```

Also to spin up a new device, there is only a simple keyword, `Setup`.

```
${DEVICE_SN}=    Setup
```

The main libraries which were developed (and hosted in an independent repo) are:

* [robotframework-c8y](https://github.com/reubenmiller/robotframework-c8y) - Provide Cumulocity IoT related actions and assertions
* [robotframework-devicelibrary](https://github.com/reubenmiller/robotframework-devicelibrary) - Device Library which is used to create an abstract test interface which sets up a "device under test" and provides an API to interactive with them. The goal of the library is to provide the same interface for docker, ssh, or on a local device (e.g. where the test code is running)

The project also provides its own `ThinEdgeIO` robot framework library which subclasses the `robotframework-devicelibrary` class. This allows thin.edge specific keywords to be added.

#### Support running integration tests against simulated devices (containers)

Being able to run the integration tests against a simulated device (e.g. a docker device), to remove the reliance on physical devices. Testing on physical devices will still be done (using the same code), however such tests will be most likely triggered on a more thorough testing cycle performed before a release. The exact specific when/how there tests are run on real devices is still to be clarified.

Some of the advantages of running tests agains a simulated device are:
* integration tests can run on the cloud hosted GitHub runner (this allows us to support more tests in parallel)
* tests can be run in parallel as each device is independent of each other
* support simulating a "network disconnect" but still allowing to communicate with the device under test

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>
#1711 

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: tests/RobotFramework/SETUP.md [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+# Setup
+
+## Pre-requisites
+
+Before you can run the tests you need to install the pre-requisites:
+
+* docker
+* python3 (>=3.10)
+* pip3
+* nodejs (>=18)
```
Comment (didier-wenzek) [inline tests/RobotFramework/SETUP.md:45]: Why is nodejs required?

Comment (reubenmiller) [reply tests/RobotFramework/SETUP.md:45]: Nodejs is still used for some parts of the tutorial settings. I guess it would be technically an optional dependency.

So I can change the working regarding it.

Comment (didier-wenzek) [reply tests/RobotFramework/SETUP.md:45]: > I guess it would be technically an optional dependency.

Don't spend too much time on that. I will try without myself.

**Comment by didier-wenzek** [comment]
Apologize: I involuntarily clicked on the new button "This branch is out-of-date with the base branch" and this merged branch 'main' into add-robot-libraries without requesting any confirmation. Better to revert this merge commit.

**Comment by reubenmiller** [comment]
> Apologize: I involuntarily clicked on the new button "This branch is out-of-date with the base branch" and this merged branch 'main' into add-robot-libraries without requesting any confirmation. Better to revert this merge commit.

I did wonder why there were two commits coming from remote (on my local machine) :D all good. It will be squashed anyway.

And that new button is actually my doing. I changed a GitHub option to show the button to users to try to encourage users to have an up-to-date state in their PRs.

File: tests/RobotFramework/SETUP.md (hunk 1 above)
Comment (reubenmiller) [reply tests/RobotFramework/SETUP.md]: I removed the note about nodejs, as is currently for an semi-automated test, but I would like to also change it as browser testing in notoriously unreliable and it does not bring much value imho.

File: tests/RobotFramework/SETUP.md [hunk 2]
```diff
@@ -0,0 +1,211 @@
+# Integration testing
```
[the remark this answers is not in the document]
Comment (reubenmiller) [reply tests/RobotFramework/SETUP.md]: I've pushed the refactored changes. Now the main setup info is in the `README.md`, and I renamed the test device info to `TEST_DEVICES.md` and simplified that page (to minimize the maintenance effort for us)

File: .github/workflows/integration-tests.yml [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+on:
+  # Use a manual approval process before PR's are given access to
+  # the secrets which are required to run the integration tests.
+  # The PR code should be manually approved to see if it can be trusted.
+  # When in doubt, do not approve the test run.
+  # Reference: https://dev.to/petrsvihlik/using-environment-protection-rules-to-secure-secrets-when-building-external-forks-with-pullrequesttarget-hci
+  pull_request_target:
+    branches: [ main, add-robot-libraries ]
+  push:
+    branches: [ main, add-robot-libraries ]
```
Comment (PradeepKiruvale) [inline .github/workflows/integration-tests.yml:11]: So, to run the integration tests one has to push it to one of these branches right? Then how they will run on every PR?

Comment (reubenmiller) [reply .github/workflows/integration-tests.yml:11]: Good catch, the `add-robot-libraries` is only activated for my PR testing (to ensure the workflow actually worked). I wanted to leave it in place until I have demoed the new testing setup to the team, then delete it before merging. I will use this conversation as a TODO.

[ ] Remove `add-robot-libraries` from the `integration-test` workflow

**Comment by reubenmiller** [comment]
Update: I discovered further problems with running the tests locally on `arm7l` due to the `cryptography` python dependency which wanted to compile Rust code as the underlying lib requires it and there are no pre-built python wheels for `arm7l`.

**TLDR;** I reworked the dependencies and removed the `poetry` package manager config
* Switched to plain python virtual environments (managed by in-built `venv` package). `poetry` was also suffering from the same problem of having a dependency to the `cryptography` package.
* Split the dependencies into different `requirement*.txt` files, where the the device can only install what is needed for the specific test device adapter (e.g. `docker`, `ssh` or `local`). This means that ssh or docker dependencies don't need to be installed if they are running the tests locally on the device (removing the need to compile the `cryptography` dependency)

**Comment by didier-wenzek** [comment]
Things are not working for me. I'm missing something, but what?

First, for the installation steps, I have had to hack a bit.

I started with a missing dependency, despite having installed the per-requisites (docker, docker-compose, python3, python3-pip): 

```shell
$ ./bin/setup.sh
The virtual environment was not created successfully because ensurepip is not
available.  On Debian/Ubuntu systems, you need to install the python3-venv
package using the following command.

    apt install python3.8-venv

You may need to use sudo with that command.  After installing the python3-venv
package, recreate your virtual environment.
```

Then, once `python3.8-venv` installed, I got an other error:

```shell
$ ./bin/setup.sh
Activating virtual environment
./bin/setup.sh: line 29: .venv/bin/activate: No such file or directory
```
Indeed, the content of that file were missing:

```
$ ls -lh .venv/bin/
total 0
lrwxrwxrwx 1 didier didier  7 févr. 15 18:08 python -> python3
lrwxrwxrwx 1 didier didier 16 févr. 15 18:08 python3 -> /usr/bin/python3
```

To move forward, the only way I found has been to delete this `.venv` directory and to relaunch the setup.
Then the installation went well with no (visible) errors, leading to ~40 files in `.venv/bin`.

But all the tests are failing with similar errors:

```
==============================================================================
Tests.Cumulocity.Firmware                                                     
==============================================================================
Tests.Cumulocity.Firmware.Firmware Operation                                  
==============================================================================
Successful firmware operation                                         | FAIL |
Setup failed:
No keyword with name 'Setup' found.

Also teardown failed:
No keyword with name 'Get Logs' found.
```

What I am missing?
* I activated the environment with `source .venv/bin/activate`
* I updated the `.env` file.

I have a question about the `.env` file. What has to be provided for `C8Y_BASEURL`: `latest.stage.c8y.io` or `didier.latest.stage.c8y.io` as I do for tedge?

**Comment by reubenmiller** [comment]
> ```shell
> venv
> ```

I think I have to add the `venv` to the pre-requisites list...but it is actually installed by default on almost all other system (except the debian it seems). The `.venv` folder seems to be only partially created, which required to delete it using `rm -rf .venv`.

You have to re-run the `./bin/setup.sh` script as this adds the `libraries` path to the python virtual environment. Though that should have worked via the `python3 bin/appendpath.py` line in the setup script...are you sure you definitely re-ran the script? If so, then maybe there is something happening there.

**Review by didier-wenzek — APPROVED** [review]
Approved, as a great step for thin-edge.

```
$ invoke build
$ invoke test
...
142 tests, 135 passed, 2 failed, 5 skipped.
===================================================
Output:  /home/didier/production/thin-edge.io/tests/RobotFramework/output/output.xml
Log:     /home/didier/production/thin-edge.io/tests/RobotFramework/output/log.html
Report:  /home/didier/production/thin-edge.io/tests/RobotFramework/output/report.html
Total testing: 35 minutes 6.19 seconds
Elapsed time:  5 minutes 14.5 seconds

```

It's so great to have the tests running without littering my laptop or my cloud tenant!
I still have to learn how to explore test outputs and add my own tests, though.
